### PR TITLE
refactor: decouple UserAccounts from NSTemplateSets

### PR DESCRIPTION
**Pull request by xcoulon**
Apply changes after switching to a pointer for `UserAccount.Spec.NSTemplateSet` 

See also:
- https://github.com/codeready-toolchain/api/pull/263
- https://github.com/codeready-toolchain/toolchain-common/pull/216

Fixes Fixes https://issues.redhat.com/browse/CRT-1305

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: controllers/usersignup/masteruserrecord.go [hunk 1]
```diff
@@ -22,7 +22,7 @@ func migrateOrFixMurIfNecessary(mur *toolchainv1alpha1.MasterUserRecord, nstempl
 			changed = true
 		}
 		nsTemplateSet := userAccount.Spec.NSTemplateSet
-		if nsTemplateSet.TierName == "" {
+		if nsTemplateSet == nil || nsTemplateSet.TierName == "" {
```
Comment (MatousJobanek) [inline controllers/usersignup/masteruserrecord.go:25]: if the `nsTemplateSet` is nil, then it should not add the default `NSTemplateSet` - we want to keep it empty so `UserAccount` controller skips it, does it make sense?

**Review by MatousJobanek — COMMENTED** [review]
Could you please add one unit test for the case when NSTemplateset is nil to the `changetierrequest_controller_test.go`?

File: controllers/templateupdaterequest/templateupdaterequest_controller.go [hunk 2]
```diff
@@ -183,11 +183,13 @@ func (r Reconciler) updateTemplateRefs(logger logr.Logger, tur toolchainv1alpha1
 			}
 			mur.Spec.UserAccounts[i] = ua
 			// also, update the tier template hash label
-			hash, err := nstemplatetier.ComputeHashForNSTemplateSetSpec(ua.Spec.NSTemplateSet)
-			if err != nil {
-				return err
+			if ua.Spec.NSTemplateSet != nil {
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/templateupdaterequest/templateupdaterequest_controller.go]: See 3f2de24 with a test, although, I'm not sure this is the behaviour we would expect (but do we really expect such a case to happen?)

File: controllers/changetierrequest/changetierrequest_controller.go [hunk 3]
```diff
@@ -186,11 +186,13 @@ func (r *Reconciler) changeTier(logger logr.Logger, changeTierRequest *toolchain
 	}
 	// then we compute again *all* hashes, in case we removed the entry for a single target cluster, but others still "use" it.
 	for _, ua := range mur.Spec.UserAccounts {
-		hash, err := nstemplatetier.ComputeHashForNSTemplateSetSpec(ua.Spec.NSTemplateSet)
-		if err != nil {
-			return r.wrapErrorWithStatusUpdate(logger, changeTierRequest, r.setStatusChangeFailed, err, "unable to compute hash for NSTemplateTier with name '%s'", nsTemplateTier.Name)
+		if ua.Spec.NSTemplateSet != nil {
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/changetierrequest/changetierrequest_controller.go]: see 0d3af3c for the removal of the superfluous check, but I'm not sure what you meant by "So you should add the check there as well"

File: controllers/usersignup/masteruserrecord.go (hunk 1 above)
Comment (xcoulon) [reply controllers/usersignup/masteruserrecord.go]: done in 34c4a4f

File: controllers/usersignup/masteruserrecord.go [hunk 4]
```diff
@@ -33,15 +33,17 @@ func migrateOrFixMurIfNecessary(mur *toolchainv1alpha1.MasterUserRecord, nstempl
 		tierName := ua.Spec.NSTemplateSet.TierName
 		// only set the label if it is missing.
 		if _, ok := mur.Labels[nstemplatetier.TemplateTierHashLabelKey(tierName)]; !ok {
-			hash, err := nstemplatetier.ComputeHashForNSTemplateSetSpec(ua.Spec.NSTemplateSet)
-			if err != nil {
-				return false, err
+			if ua.Spec.NSTemplateSet != nil {
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/usersignup/masteruserrecord.go]: done in https://github.com/codeready-toolchain/host-operator/commit/34c4a4f46b91106154477f8212d2fd09e95eecc4 and 05b5d0e

**Comment by xcoulon** [comment]
> Could you please add one unit test for the case when NSTemplateset is nil to
> 

sure 
> * `changetierrequest_controller_test.go`

done in 91fe4f9
> * `templateupdaterequest_controller_test.go`

done in https://github.com/codeready-toolchain/host-operator/commit/3f2de247c8d57f9171c1bb627732bf1c1bc5d14f

File: controllers/changetierrequest/changetierrequest_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/changetierrequest/changetierrequest_controller.go]: actually, I reverted this change above and added another nil-check as you suggested: 91fe4f9

**Review by MatousJobanek — APPROVED** [review]
Awesome, thanks a lot for taking the extra time to address my comments :+1:
